File: algorithms/algorithms-python/leetcode/Question_0024_Swap_Nodes_in_Pairs.py
#!/usr/bin/python
# coding: utf-8

# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head:
            return head
        help = ListNode(0)
        help.next = head
        pre, cur, next = help, head, None
        while cur and cur.next:
            next = cur.next.next
            pre.next = cur.next
            cur.next.next = cur
            cur.next = next
            pre = cur
            cur = next
        return help.next

File: algorithms/algorithms-cplusplus/leetcode/Question_0024_Swap_Nodes_in_Pairs.cc

// Definition for singly-linked list.
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
 public:
  ListNode* swapPairs(ListNode* head) {
    ListNode dummy(0);
    dummy.next = head;
    ListNode *p = &dummy, *next;
    while (p->next && p->next->next) {
      next = p->next->next;
      p->next->next = next->next;
      next->next = p->next;
      p->next = next;
      p = p->next->next;
    }
    return dummy.next;
  }
};
